Add support for Abstraction data types in enter_block()

yimjay opened the pull request:
Define a new block to specify Abstraction data types and configure enter_block() to recognize the block.

bnestere commented on lib/analysis.pt:
```diff
@@ -7,6 +7,7 @@ include analysis.pi
    case CODE.FunctionDecl: newtab=block[FunctionDecl.symtab];
    case CODE.StmtBlock: newtab=block[StmtBlock.symtab];
    case CODE.DeclarationBlock: newtab=block[DeclarationBlock.symtab];
+   case CODE.AbstractionDataType: newtab=block[AbstractionDataType.symtab];
```
Updating POET to account for the Abstraction type doesn't seem right to me.. Is there a way this can be generalized to facilitate more of a general approach that allows any type to have a symbol table? That would be more extensible so for any project that wants to use a symbol table for their own code template, they don't have to extend POET itself to allow it.

yimjay replied:
A more general approach was implemented: the enter_block() function now accepts tables as inputs

qingyi-yan commented on lib/analysis.pt:
```diff
@@ -7,6 +7,7 @@ include analysis.pi
    case CODE.FunctionDecl: newtab=block[FunctionDecl.symtab];
    case CODE.StmtBlock: newtab=block[StmtBlock.symtab];
    case CODE.DeclarationBlock: newtab=block[DeclarationBlock.symtab];
+   case CODE.AbstractionBlock: newtab=block[AbstractionBlock.symtab];
```
change to CODE.Abstraction ?

Also, I think we should add another case where block itself is the symbol table of a new block. to address Brandon's comment.

case MAP: new_tab = block;

yimjay commented:
Added support for a block itself in [last commit](https://github.com/qingyi-yan/POET/pull/9/commits/dc81d98f2910e88c00ba5fb6655bbce4e758f01f) pushed on 8/12. 
* Added `case MAP : newtab = block;` to `XFORM.enter_block()` in `analysis.pt`

Tested by running the following:
```
testMap = MAP{"a" => 1,
                "b" => 2,
                "c" => 3};
XFORM.enter_block(testMap);
DEBUG{GLOBAL.SymTable};
XFORM.exit_block(testMap);
```

where output was:
```
DEBUGGING at level 1
END DEBUGGING at level 1
DEBUGGING result: SymTable===>POETList:(MAP{"a"=>1;"b"=>2;"c"=>3;})
```

bnestere reviewed (approved):
Looks good, thanks!
